security/remote_devices/docker-swarm/docker-stack-edgex.yml

```
@@ -0,0 +1,617 @@
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/edgex-support-scheduler/
+  edgex-core-command:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/edgex-core-command/
+
+services:
+  consul:
+    image: edgexfoundry/docker-edgex-consul:1.2.0
+    ports:
+      - "8400:8400"
+      - "8500:8500"
+    hostname: edgex-core-consul
+    networks:
+      edgex-network:
+        aliases:
+            - edgex-core-consul
+    volumes:
+      - consul-config:/consul/config:z
+      - consul-data:/consul/data:z
+      - consul-scripts:/consul/scripts:z
+      - secrets-ca-volume:/tmp/edgex/secrets/ca:ro,z
+      - edgex-consul:/tmp/edgex/secrets/edgex-consul:ro,z
+      - edgex-vault:/tmp/edgex/secrets/edgex-vault:ro,z
+      - edgex-kong:/tmp/edgex/secrets/edgex-kong:ro,z
+    environment:
+      - "SECRETSTORE_SETUP_DONE_FLAG=/tmp/edgex/secrets/edgex-consul/.secretstore-setup-done"
+      - EDGEX_DB=redis
+      - EDGEX_SECURE=true
+      - CONSUL_BIND_INTERFACE=eth0
+    depends_on:
+      - security-secrets-setup
+    deploy: 
+      placement: 
+        constraints: 
+          - node.hostname != worker-node 
+
+  vault:
+    image: vault:1.3.1
+    hostname: edgex-vault
+    networks:
+      edgex-network:
+        aliases:
+            - edgex-vault
+    ports:
+      - "8200:8200"
+    # cap_add not allowed in docker swarm, this is a security issue and I don't recommend disabling this in production
+    # cap_add:
+    #   - "IPC_LOCK"
+    tmpfs:
+      - /vault/config
+    entrypoint: ["/vault/init/start_vault.sh"]
+    environment:
+      - VAULT_ADDR=https://edgex-vault:8200
+      - VAULT_CONFIG_DIR=/vault/config
+      - VAULT_UI=true
+      - SKIP_SETCAP=true
+      - |
+        VAULT_LOCAL_CONFIG=
+          listener "tcp" { 
+              address = "edgex-vault:8200" 
+              tls_disable = "0" 
+              cluster_address = "edgex-vault:8201" 
+              tls_min_version = "tls12" 
+              tls_client_ca_file ="/tmp/edgex/secrets/edgex-vault/ca.pem" 
+              tls_cert_file ="/tmp/edgex/secrets/edgex-vault/server.crt" 
+              tls_key_file = "/tmp/edgex/secrets/edgex-vault/server.key" 
+              tls_perfer_server_cipher_suites = "true"
+          } 
+          backend "consul" { 
+              path = "vault/" 
+              address = "edgex-core-consul:8500" 
+              scheme = "http" 
+              redirect_addr = "https://edgex-vault:8200" 
+              cluster_addr = "https://edgex-vault:8201" 
+          } 
+          default_lease_ttl = "168h" 
+          max_lease_ttl = "720h"
+          disable_mlock = "true"
+    volumes:
+      - vault-file:/vault/file:z
+      - vault-logs:/vault/logs:z
+      - vault-init:/vault/init:ro,z
+      - edgex-vault:/tmp/edgex/secrets/edgex-vault:ro,z
+    depends_on:
+      - consul
+      - security-secrets-setup
+    deploy: 
+      placement: 
```

```suggestion
      endpoint_mode: dnsrr
      placement:
```

Without this change to endpoint_mode, the DNS lookup for edgex-vault resolves to the service's virtual IP, which is not bindable in the container.